File: backend/internal/permissions/permissions.go
package permissions

import (
	"encoding/json"
)

const (
	LeadScopeAll                 = "all"
	LeadScopeAssigned            = "assigned"
	LeadScopeFollowed            = "followed"
	LeadScopeAssignedAndFollowed = "assigned_and_followed"

	ActionSettingsAdmin      = "settings_admin"
	ActionPipelinesRouting   = "pipelines_routing"
	ActionContractsPartners  = "contracts_partners"
	ActionBilling            = "billing"
	ActionAppointmentsManage = "appointments_manage"
)

const (
	NavDashboard     = "dashboard"
	NavLeads         = "leads"
	NavFields        = "fields"
	NavAppointments  = "appointments"
	NavCalendars     = "calendars"
	NavCalls         = "calls"
	NavBoard         = "board"
	NavPipelines     = "pipelines"
	NavBuyers        = "buyers"
	NavPublishers    = "publishers"
	NavContracts     = "contracts"
	NavCollaboration = "collaboration"
	NavSources       = "sources"
	NavWebhooks      = "webhooks"
	NavRouting       = "routing"
	NavLogs          = "logs"
	NavRoutes        = "routes"
	NavSettings      = "settings"
	NavBilling       = "billing"
	NavIntegrations  = "integrations"
)

var allActionKeys = []string{
	ActionSettingsAdmin, ActionPipelinesRouting, ActionContractsPartners,
	ActionBilling, ActionAppointmentsManage,
}

type Actions struct {
	SettingsAdmin      bool `json:"settings_admin"`
	PipelinesRouting   bool `json:"pipelines_routing"`
	ContractsPartners  bool `json:"contracts_partners"`
	Billing            bool `json:"billing"`
	AppointmentsManage bool `json:"appointments_manage"`
}

type Overrides struct {
	Nav       map[string]bool `json:"nav,omitempty"`
	LeadScope string          `json:"lead_scope,omitempty"`
	Actions   map[string]bool `json:"actions,omitempty"`
}

type Effective struct {
	Nav       map[string]bool `json:"nav"`
	LeadScope string          `json:"lead_scope"`
	Actions   Actions         `json:"actions"`
}

func (e Effective) CanNav(key string) bool {
	return e.Nav != nil && e.Nav[key]
}

func (e Effective) CanAction(key string) bool {
	switch key {
	case ActionSettingsAdmin:
		return e.Actions.SettingsAdmin
	case ActionPipelinesRouting:
		return e.Actions.PipelinesRouting
	case ActionContractsPartners:
		return e.Actions.ContractsPartners
	case ActionBilling:
		return e.Actions.Billing
	case ActionAppointmentsManage:
		return e.Actions.AppointmentsManage
	default:
		return false
	}
}

func (e Effective) IsFullAdmin() bool {
	if e.LeadScope != LeadScopeAll {
		return false
	}
	if !e.Actions.SettingsAdmin || !e.Actions.PipelinesRouting || !e.Actions.ContractsPartners ||
		!e.Actions.Billing || !e.Actions.AppointmentsManage {
		return false
	}
	for _, v := range e.Nav {
		if !v {
			return false
		}
	}
	return true
}

func ParseOverrides(raw []byte) Overrides {
	if len(raw) == 0 {
		return Overrides{}
	}
	var o Overrides
	_ = json.Unmarshal(raw, &o)
	return o
}

func MarshalOverrides(o Overrides) ([]byte, error) {
	if o.LeadScope == "" && len(o.Nav) == 0 && len(o.Actions) == 0 {
		return []byte("{}"), nil
	}
	return json.Marshal(o)
}

func FullAccess(accountType string) Effective {
	return presetForRole("admin", accountType)
}

func Resolve(role, accountType string, raw []byte) Effective {
	return merge(presetForRole(role, accountType), ParseOverrides(raw))
}

func PresetForRole(role, accountType string) Effective {
	return presetForRole(role, accountType)
}

func Delta(role, accountType string, effective Effective) Overrides {
	preset := presetForRole(role, accountType)
	var o Overrides
	if effective.LeadScope != preset.LeadScope {
		o.LeadScope = effective.LeadScope
	}
	if nav := diffBoolMap(preset.Nav, effective.Nav); len(nav) > 0 {
		o.Nav = nav
	}
	if actions := diffActionMaps(preset.Actions, effective.Actions); len(actions) > 0 {
		o.Actions = actions
	}
	return o
}

func merge(preset Effective, o Overrides) Effective {
	out := preset
	if o.LeadScope != "" {
		out.LeadScope = o.LeadScope
	}
	if len(o.Nav) > 0 {
		out.Nav = applyBoolMap(out.Nav, o.Nav)
	}
	if len(o.Actions) > 0 {
		out.Actions = applyActions(out.Actions, o.Actions)
	}
	return out
}

func presetForRole(role, accountType string) Effective {
	nav := defaultNavForAccount(accountType)
	actions := Actions{}
	leadScope := LeadScopeAssigned

	switch role {
	case "admin":
		leadScope = LeadScopeAll
		actions = Actions{
			SettingsAdmin: true, PipelinesRouting: true, ContractsPartners: true,
			Billing: true, AppointmentsManage: true,
		}
		for k := range nav {
			nav[k] = true
		}
	case "follower":
		leadScope = LeadScopeFollowed
		for k := range nav {
			if !adminOnlyNav(k, accountType) {
				nav[k] = true
			}
		}
	default:
		for k := range nav {
			if !adminOnlyNav(k, accountType) {
				nav[k] = true
			}
		}
	}
	return Effective{Nav: nav, LeadScope: leadScope, Actions: actions}
}

func defaultNavForAccount(accountType string) map[string]bool {
	keys := []string{
		NavDashboard, NavLeads, NavFields, NavAppointments, NavCalendars, NavCalls,
		NavBoard, NavPipelines, NavContracts, NavCollaboration, NavWebhooks,
		NavSettings, NavBilling, NavIntegrations,
	}
	switch accountType {
	case "publisher":
		keys = append(keys, NavBuyers, NavSources, NavRouting, NavLogs)
	case "buyer":
		keys = append(keys, NavPublishers, NavRoutes, NavLogs)
	}
	nav := make(map[string]bool, len(keys))
	for _, k := range keys {
		nav[k] = false
	}
	return nav
}

func adminOnlyNav(key, accountType string) bool {
	switch key {
	case NavFields, NavPipelines, NavCollaboration, NavWebhooks, NavIntegrations:
		return true
	case NavBuyers, NavSources, NavRouting:
		return accountType == "publisher"
	case NavLogs:
		return true
	default:
		return false
	}
}

func applyBoolMap(base, patch map[string]bool) map[string]bool {
	out := make(map[string]bool, len(base))
	for k, v := range base {
		out[k] = v
	}
	for k, v := range patch {
		if _, ok := out[k]; ok {
			out[k] = v
		}
	}
	return out
}

func diffBoolMap(preset, effective map[string]bool) map[string]bool {
	out := map[string]bool{}
	for k, pv := range preset {
		if ev, ok := effective[k]; ok && ev != pv {
			out[k] = ev
		}
	}
	return out
}

func actionsToMap(a Actions) map[string]bool {
	return map[string]bool{
		ActionSettingsAdmin:      a.SettingsAdmin,
		ActionPipelinesRouting:   a.PipelinesRouting,
		ActionContractsPartners:  a.ContractsPartners,
		ActionBilling:            a.Billing,
		ActionAppointmentsManage: a.AppointmentsManage,
	}
}

func mapToActions(m map[string]bool) Actions {
	return Actions{
		SettingsAdmin:      m[ActionSettingsAdmin],
		PipelinesRouting:   m[ActionPipelinesRouting],
		ContractsPartners:  m[ActionContractsPartners],
		Billing:            m[ActionBilling],
		AppointmentsManage: m[ActionAppointmentsManage],
	}
}

func applyActions(base Actions, patch map[string]bool) Actions {
	m := actionsToMap(base)
	for k, v := range patch {
		if _, ok := m[k]; ok {
			m[k] = v
		}
	}
	return mapToActions(m)
}

func diffActionMaps(preset Actions, effective Actions) map[string]bool {
	pm := actionsToMap(preset)
	em := actionsToMap(effective)
	return diffBoolMap(pm, em)
}

func HasAssignedScope(scope string) bool {
	return scope == LeadScopeAll || scope == LeadScopeAssigned || scope == LeadScopeAssignedAndFollowed
}

func HasFollowedScope(scope string) bool {
	return scope == LeadScopeFollowed || scope == LeadScopeAssignedAndFollowed
}

func IsFollowedOnly(scope string) bool {
	return scope == LeadScopeFollowed
}

func ValidateOverrides(accountType string, o Overrides) error {
	if o.LeadScope != "" && o.LeadScope != LeadScopeAll && o.LeadScope != LeadScopeAssigned &&
		o.LeadScope != LeadScopeFollowed && o.LeadScope != LeadScopeAssignedAndFollowed {
		return errInvalid("invalid lead_scope")
	}
	navKeys := defaultNavForAccount(accountType)
	for k := range o.Nav {
		if _, ok := navKeys[k]; !ok {
			return errInvalid("unknown nav key: " + k)
		}
	}
	for k := range o.Actions {
		valid := false
		for _, ak := range allActionKeys {
			if k == ak {
				valid = true
				break
			}
		}
		if !valid {
			return errInvalid("unknown action key: " + k)
		}
	}
	return nil
}

type validationError string

func (e validationError) Error() string { return string(e) }

func errInvalid(msg string) error { return validationError(msg) }

func ToMap(e Effective) map[string]any {
	return map[string]any{
		"nav":        e.Nav,
		"lead_scope": e.LeadScope,
		"actions": map[string]bool{
			ActionSettingsAdmin:      e.Actions.SettingsAdmin,
			ActionPipelinesRouting:   e.Actions.PipelinesRouting,
			ActionContractsPartners:  e.Actions.ContractsPartners,
			ActionBilling:            e.Actions.Billing,
			ActionAppointmentsManage: e.Actions.AppointmentsManage,
		},
	}
}

File: frontend/src/lib/permissions.ts
import type { AccountType, CurrentUser, Role } from "@/types";

export const LeadScopeAll = "all";
export const LeadScopeAssigned = "assigned";
export const LeadScopeFollowed = "followed";
export const LeadScopeAssignedAndFollowed = "assigned_and_followed";

export const ActionSettingsAdmin = "settings_admin";
export const ActionPipelinesRouting = "pipelines_routing";
export const ActionContractsPartners = "contracts_partners";
export const ActionBilling = "billing";
export const ActionAppointmentsManage = "appointments_manage";

export const NavDashboard = "dashboard";
export const NavLeads = "leads";
export const NavFields = "fields";
export const NavAppointments = "appointments";
export const NavCalendars = "calendars";
export const NavCalls = "calls";
export const NavBoard = "board";
export const NavPipelines = "pipelines";
export const NavBuyers = "buyers";
export const NavPublishers = "publishers";
export const NavContracts = "contracts";
export const NavCollaboration = "collaboration";
export const NavSources = "sources";
export const NavWebhooks = "webhooks";
export const NavRouting = "routing";
export const NavLogs = "logs";
export const NavRoutes = "routes";
export const NavSettings = "settings";
export const NavBilling = "billing";
export const NavIntegrations = "integrations";

export type LeadScope =
  | typeof LeadScopeAll
  | typeof LeadScopeAssigned
  | typeof LeadScopeFollowed
  | typeof LeadScopeAssignedAndFollowed;

export type ActionKey =
  | typeof ActionSettingsAdmin
  | typeof ActionPipelinesRouting
  | typeof ActionContractsPartners
  | typeof ActionBilling
  | typeof ActionAppointmentsManage;

export interface EffectivePermissions {
  nav: Record<string, boolean>;
  lead_scope: LeadScope;
  actions: Record<ActionKey, boolean>;
}

export interface PermissionOverrides {
  nav?: Record<string, boolean>;
  lead_scope?: LeadScope;
  actions?: Partial<Record<ActionKey, boolean>>;
}

function adminOnlyNav(key: string, accountType: AccountType): boolean {
  switch (key) {
    case NavFields:
    case NavPipelines:
    case NavCollaboration:
    case NavWebhooks:
    case NavIntegrations:
      return true;
    case NavBuyers:
    case NavSources:
    case NavRouting:
      return accountType === "publisher";
    case NavLogs:
      return true;
    default:
      return false;
  }
}

function defaultNavKeys(accountType: AccountType): string[] {
  const keys = [
    NavDashboard, NavLeads, NavFields, NavAppointments, NavCalendars, NavCalls,
    NavBoard, NavPipelines, NavContracts, NavCollaboration, NavWebhooks,
    NavSettings, NavBilling, NavIntegrations,
  ];
  if (accountType === "publisher") return [...keys, NavBuyers, NavSources, NavRouting, NavLogs];
  if (accountType === "buyer") return [...keys, NavPublishers, NavRoutes, NavLogs];
  return keys;
}

export function presetForRole(role: Role, accountType: AccountType): EffectivePermissions {
  const nav: Record<string, boolean> = {};
  for (const k of defaultNavKeys(accountType)) nav[k] = false;

  const actions: Record<ActionKey, boolean> = {
    settings_admin: false,
    pipelines_routing: false,
    contracts_partners: false,
    billing: false,
    appointments_manage: false,
  };

  let lead_scope: LeadScope = LeadScopeAssigned;

  if (role === "admin") {
    lead_scope = LeadScopeAll;
    for (const k of Object.keys(nav)) nav[k] = true;
    for (const k of Object.keys(actions) as ActionKey[]) actions[k] = true;
  } else if (role === "follower") {
    lead_scope = LeadScopeFollowed;
    for (const k of Object.keys(nav)) {
      if (!adminOnlyNav(k, accountType)) nav[k] = true;
    }
  } else {
    for (const k of Object.keys(nav)) {
      if (!adminOnlyNav(k, accountType)) nav[k] = true;
    }
  }

  return { nav, lead_scope, actions };
}

export function resolvePermissions(
  role: Role,
  accountType: AccountType,
  overrides?: PermissionOverrides | Record<string, unknown>
): EffectivePermissions {
  const preset = presetForRole(role, accountType);
  if (!overrides || Object.keys(overrides).length === 0) return preset;

  const o = overrides as PermissionOverrides;
  const out = { ...preset, nav: { ...preset.nav }, actions: { ...preset.actions } };
  if (o.lead_scope) out.lead_scope = o.lead_scope;
  if (o.nav) {
    for (const [k, v] of Object.entries(o.nav)) {
      if (k in out.nav) out.nav[k] = v;
    }
  }
  if (o.actions) {
    for (const [k, v] of Object.entries(o.actions)) {
      if (k in out.actions) out.actions[k as ActionKey] = v;
    }
  }
  return out;
}

export function effectivePermissions(user: CurrentUser | null | undefined): EffectivePermissions | null {
  if (!user?.effective_permissions) return null;
  const ep = user.effective_permissions;
  return {
    nav: ep.nav ?? {},
    lead_scope: (ep.lead_scope as LeadScope) ?? LeadScopeAssigned,
    actions: {
      settings_admin: !!ep.actions?.settings_admin,
      pipelines_routing: !!ep.actions?.pipelines_routing,
      contracts_partners: !!ep.actions?.contracts_partners,
      billing: !!ep.actions?.billing,
      appointments_manage: !!ep.actions?.appointments_manage,
    },
  };
}

export function canNav(user: CurrentUser | null | undefined, key: string): boolean {
  if (!user) return false;
  if (user.account_type === "platform") return true;
  if (user.impersonating || user.is_switched) return true;
  const ep = effectivePermissions(user);
  return ep?.nav[key] ?? presetForRole(user.role, user.account_type).nav[key] ?? false;
}

export function canAction(user: CurrentUser | null | undefined, key: ActionKey): boolean {
  if (!user) return false;
  if (user.account_type === "platform") return true;
  if (user.impersonating || user.is_switched) return true;
  const ep = effectivePermissions(user);
  return ep?.actions[key] ?? false;
}

export function leadScope(user: CurrentUser | null | undefined): LeadScope {
  if (!user) return LeadScopeAssigned;
  if (user.impersonating || user.is_switched) return LeadScopeAll;
  const ep = effectivePermissions(user);
  return ep?.lead_scope ?? presetForRole(user.role, user.account_type).lead_scope;
}

export function leadScopeToFlags(scope: LeadScope): { assigned: boolean; followed: boolean } {
  return {
    assigned: scope === LeadScopeAssigned || scope === LeadScopeAssignedAndFollowed,
    followed: scope === LeadScopeFollowed || scope === LeadScopeAssignedAndFollowed,
  };
}

export function leadScopeFromFlags(
  assigned: boolean,
  followed: boolean,
  role: Role
): LeadScope | null {
  if (assigned && followed) return LeadScopeAssignedAndFollowed;
  if (assigned) return LeadScopeAssigned;
  if (followed) return LeadScopeFollowed;
  if (role === "admin") return LeadScopeAll;
  return null;
}

export function isValidLeadVisibility(role: Role, scope: LeadScope): boolean {
  if (scope === LeadScopeAll) return role === "admin";
  return scope === LeadScopeAssigned || scope === LeadScopeFollowed || scope === LeadScopeAssignedAndFollowed;
}

export function canCreateLead(user: CurrentUser | null | undefined): boolean {
  return leadScope(user) !== LeadScopeFollowed;
}

export function canEditLead(
  user: CurrentUser | null | undefined,
  lead: { assigned_user_id?: number | null }
): boolean {
  const scope = leadScope(user);
  if (scope === LeadScopeAll) return true;
  if (scope === LeadScopeFollowed) return false;
  return lead.assigned_user_id === Number(user?.id);
}

export function canSeeAllLeads(user: CurrentUser | null | undefined): boolean {
  return leadScope(user) === LeadScopeAll;
}

export function deltaFromEffective(
  role: Role,
  accountType: AccountType,
  effective: EffectivePermissions
): PermissionOverrides {
  const preset = presetForRole(role, accountType);
  const overrides: PermissionOverrides = {};
  if (effective.lead_scope !== preset.lead_scope) overrides.lead_scope = effective.lead_scope;

  const nav: Record<string, boolean> = {};
  for (const k of Object.keys(preset.nav)) {
    if (effective.nav[k] !== preset.nav[k]) nav[k] = effective.nav[k];
  }
  if (Object.keys(nav).length > 0) overrides.nav = nav;

  const actions: Partial<Record<ActionKey, boolean>> = {};
  for (const k of Object.keys(preset.actions) as ActionKey[]) {
    if (effective.actions[k] !== preset.actions[k]) actions[k] = effective.actions[k];
  }
  if (Object.keys(actions).length > 0) overrides.actions = actions;

  return overrides;
}

export interface NavSection {
  label: string;
  keys: { key: string; label: string }[];
}

export function navSections(accountType: AccountType): NavSection[] {
  const leads = [
    { key: NavLeads, label: "Leads" },
    { key: NavFields, label: "Custom Fields" },
  ];
  const appointments = [
    { key: NavAppointments, label: "Appointments" },
    { key: NavCalendars, label: "Calendars" },
  ];
  const pipeline = [
    { key: NavBoard, label: "Pipeline" },
    { key: NavPipelines, label: "Pipelines" },
  ];
  const settings = [
    { key: NavSettings, label: "Settings" },
    { key: NavBilling, label: "Billing" },
    { key: NavIntegrations, label: "Integrations" },
  ];

  if (accountType === "publisher") {
    return [
      { label: "Overview", keys: [{ key: NavDashboard, label: "Dashboard" }] },
      { label: "Leads", keys: leads },
      { label: "Appointments", keys: appointments },
      { label: "Calls", keys: [{ key: NavCalls, label: "Calls" }] },
      { label: "Pipeline", keys: pipeline },
      {
        label: "Buyers",
        keys: [
          { key: NavBuyers, label: "Buyers" },
          { key: NavContracts, label: "Contracts" },
          { key: NavCollaboration, label: "Collaboration" },
        ],
      },
      {
        label: "Routing",
        keys: [
          { key: NavSources, label: "Sources" },
          { key: NavWebhooks, label: "Webhooks" },
          { key: NavRouting, label: "Routing" },
          { key: NavLogs, label: "Logs" },
        ],
      },
      { label: "Settings", keys: settings },
    ];
  }

  return [
    { label: "Overview", keys: [{ key: NavDashboard, label: "Dashboard" }] },
    { label: "Leads", keys: leads },
    { label: "Appointments", keys: appointments },
    { label: "Calls", keys: [{ key: NavCalls, label: "Calls" }] },
    { label: "Pipeline", keys: pipeline },
    {
      label: "Publishers",
      keys: [
        { key: NavPublishers, label: "Publishers" },
        { key: NavContracts, label: "Contracts" },
        { key: NavCollaboration, label: "Collaboration" },
      ],
    },
    {
      label: "Routing",
      keys: [
        { key: NavRoutes, label: "Routes" },
        { key: NavWebhooks, label: "Webhooks" },
        { key: NavLogs, label: "Logs" },
      ],
    },
    { label: "Settings", keys: settings },
  ];
}

export const ACTION_OPTIONS: { key: ActionKey; label: string }[] = [
  { key: ActionSettingsAdmin, label: "Manage users, API keys & business profile" },
  { key: ActionPipelinesRouting, label: "Manage pipelines, fields & routing" },
  { key: ActionContractsPartners, label: "Manage contracts & partners" },
  { key: ActionBilling, label: "Manage billing" },
  { key: ActionAppointmentsManage, label: "Manage calendars & booking slots" },
];
